File: vds-promotion/src/main/java/com/ttcn/promotionsdk/ui/theme/PromotionThemeDefaults.kt
package com.ttcn.promotionsdk.ui.theme

import android.content.Context
import android.graphics.Color
import androidx.annotation.ColorInt
import com.ttcn.promotionsdk.R
import com.ttcn.promotionsdk.ui.utils.enum.PRMCoreButtonSize
import com.ttcn.promotionsdk.ui.utils.enum.PRMCoreButtonType
import com.ttcn.promotionsdk.ui.utils.enum.PRMShadowType
import com.ttcn.promotionsdk.ui.utils.enum.PRMSearchType
import com.ttcn.promotionsdk.ui.utils.extension.retrieveColor

/**
 * Internal SDK defaults extracted from component source / resources.
 * Used by theme preview and when resolving unset token properties.
 */
internal object PromotionThemeDefaults {

    /**
     * Defaults aligned with [PRMButton] PRIMARY + LARGE (common CTA in promotion screens).
     * Background is gradient in SDK; token stores gradient end color [R.color.tokenRainbowRedEnd].
     * Corner radius matches [PRMShadowType.TokenShadowsButtonLarge] (in dp).
     */
    fun button(context: Context) = ButtonToken(
        backgroundColor = context.retrieveColor(R.color.tokenRainbowRedEnd),
        textColor = context.retrieveColor(PRMCoreButtonType.PRIMARY.textColorRes),
        shadowColor = context.retrieveColor(R.color.tokenShadowsButtonColor),
        cornerRadius = buttonCornerRadiusDp(context, PRMCoreButtonSize.LARGE),
    )

    fun buttonCornerRadiusDp(context: Context, size: PRMCoreButtonSize = PRMCoreButtonSize.LARGE): Float {
        val shadowType = when (size) {
            PRMCoreButtonSize.SMALL -> PRMShadowType.TokenShadowsButtonSmall
            PRMCoreButtonSize.MEDIUM -> PRMShadowType.TokenShadowsButtonMedium
            PRMCoreButtonSize.LARGE -> PRMShadowType.TokenShadowsButtonLarge
        }
        return pxToDp(context, context.resources.getDimension(shadowType.cornerRes))
    }

    fun searchBar(context: Context) = SearchBarToken(
        borderColor = context.retrieveColor(R.color.tokenDark10),
        hintTextColor = context.retrieveColor(PRMSearchType.BASIC.hintColorRes),
        textColor = context.retrieveColor(PRMSearchType.BASIC.textColorRes),
        iconColor = context.retrieveColor(R.color.tokenDark40),
        cornerRadius = pxToDp(context, context.resources.getDimension(R.dimen.tokenBorderRadius08)),
    )

    fun listItem(context: Context) = ListItemToken(
        linkTextColor = context.retrieveColor(R.color.color_EE0033),
        usedBadgeTextColor = context.retrieveColor(R.color.tokenDark100),
        usedBadgeBackgroundColor = context.retrieveColor(R.color.color_f4f4f4),
        radioButtonStrokeColor = context.retrieveColor(R.color.tokenDark40),
        radioButtonSelectedStrokeColor = context.retrieveColor(R.color.color_EE0033),
    )

    fun tabChip(context: Context) = TabChipToken(
        activeBackgroundColor = context.retrieveColor(R.color.color_4e4e4e),
        inactiveBackgroundColor = context.retrieveColor(R.color.color_f4f4f4),
        activeTextColor = context.retrieveColor(R.color.white),
        inactiveTextColor = context.retrieveColor(R.color.color_7A7A7A),
        cornerRadius = pxToDp(context, context.resources.getDimension(R.dimen.view_size_7)),
    )

    fun tabUnderline(context: Context) = TabUnderlineToken(
        indicatorColor = context.retrieveColor(R.color.color_red_EE0033),
        activeTextColor = context.retrieveColor(R.color.black),
        inactiveTextColor = context.retrieveColor(R.color.color_7a7a7a),
        backgroundColor = context.retrieveColor(R.color.color_FBFBFB),
    )

    fun discountBadge(context: Context) = DiscountBadgeToken(
        availableTextColor = context.retrieveColor(R.color.tokenPineBlue100),
        unavailableTextColor = context.retrieveColor(R.color.color_7A7A7A),
        availableBackgroundColor = context.retrieveColor(R.color.tokenPineBlue10),
        unavailableBackgroundColor = context.retrieveColor(R.color.color_f4f4f4),
        actionTextColor = context.retrieveColor(R.color.color_EE0033),
    )

    fun defaultConfig(context: Context) = PromotionThemeConfig(
        buttonToken = button(context),
        searchBarToken = searchBar(context),
        listItemToken = listItem(context),
        tabChipToken = tabChip(context),
        tabUnderlineToken = tabUnderline(context),
        discountBadgeToken = discountBadge(context),
    )

    @ColorInt
    fun colorToHex(@ColorInt color: Int): String =
        if (Color.alpha(color) != 0xFF) {
            String.format("#%08X", color)
        } else {
            String.format("#%06X", 0xFFFFFF and color)
        }

    fun pxToDp(context: Context, px: Float): Float =
        px / context.resources.displayMetrics.density
}

File: iosPromotionUI/PromotionSDK/Theme/PromotionThemeDefaults.swift
//
//  PromotionThemeDefaults.swift
//  PromotionSDK
//
//  Giá trị mặc định **thật** của SDK, đọc từ nhánh `nil` của các component CoreUI/PromotionUI.
//  Đối ứng `PromotionThemeDefaults.kt` bên Android (nơi giá trị đến từ `R.color.*`).
//
//  Dùng cho màn cấu hình theme của host: token nào chưa set thì hiện giá trị SDK đang dùng.
//  Trước đây `ThemePlaygroundSupport.swift` của app demo phải **đoán** ("Màu default xấp xỉ SDK").
//
//  ⚠️ Giá trị ở đây **không** trùng Android ở mọi field — hai nền tảng vốn có default khác nhau
//  (badge giảm giá: teal bên Android, đỏ san hô bên iOS). Xem bảng trong `docs/Theming.md`.
//  File này **báo cáo trung thực** những gì iOS đang render, không phải sao chép Android.
//

import UIKit
@_implementationOnly import CoreUI

public enum PromotionThemeDefaults {

    /// Nút chính (`VDSButton`, size `.large`, không outline).
    ///
    /// - `backgroundColor`: SDK vẽ **gradient** `tokenViettelPayRed80 → tokenViettelPayRed100`;
    ///   token chỉ giữ màu cuối, đúng như Android giữ `tokenRainbowRedEnd`.
    /// - `cornerRadius`: SDK dùng `height / 2` (viên thuốc). Với `largeHeight = 48` thì bằng 24.
    /// - `shadowColor`: `Shadows.tokenShadowButtonLarge` là đỏ + opacity 0.25 → `#40EE0033`.
    public static var button: ButtonToken {
        ButtonToken(
            backgroundColor: Colors.tokenViettelPayRed100,
            textColor: Colors.tokenWhite,
            shadowColor: Colors.tokenViettelPayRed100.withAlphaComponent(0.25),
            cornerRadius: 24
        )
    }

    /// Ô tìm kiếm (`VDSSearchTextField`, không phải kiểu navigation).
    /// `iconColor` không có default thật — SDK render ảnh gốc, không tint. Lấy `tokenDark40` cho
    /// khớp Android và để màn preview có gì đó hiển thị.
    public static var searchBar: SearchBarToken {
        SearchBarToken(
            borderColor: Colors.tokenDark10,
            hintTextColor: Colors.tokenDark40,
            textColor: Colors.tokenDark100,
            iconColor: Colors.tokenDark40,
            cornerRadius: Sizing.tokenSizing08
        )
    }

    /// Item voucher (`PromotionCardView`).
    /// `linkTextColor` và hai màu radio **không có default trong code** — SDK giữ màu của XIB và ảnh
    /// asset gốc. Giá trị dưới đây là những gì asset đang thể hiện, để preview không trống.
    public static var listItem: ListItemToken {
        ListItemToken(
            linkTextColor: Colors.tokenViettelPayRed100,
            usedBadgeTextColor: Colors.tokenDark60,
            usedBadgeBackgroundColor: Colors.tokenDark05,
            radioButtonStrokeColor: Colors.tokenDark40,
            radioButtonSelectedStrokeColor: Colors.tokenViettelPayRed100
        )
    }

    /// Tab dạng chip (`PromotionTabView`).
    public static var tabChip: TabChipToken {
        TabChipToken(
            activeBackgroundColor: Colors.tokenDark80,
            inactiveBackgroundColor: Colors.tokenDark05,
            activeTextColor: Colors.tokenWhite,
            inactiveTextColor: Colors.tokenDark60,
            cornerRadius: 8
        )
    }

    /// Tab gạch chân (`UnderlinedSegmentControlItem`).
    /// `backgroundColor` không có default — thanh tab chỉ đổi nền khi host set token.
    public static var tabUnderline: TabUnderlineToken {
        TabUnderlineToken(
            indicatorColor: Colors.tokenRed100,
            activeTextColor: Colors.tokenDark100,
            inactiveTextColor: Colors.tokenDark60,
            backgroundColor: nil
        )
    }

    /// Badge giảm giá trên widget (`PRMEndowView`).
    public static var discountBadge: DiscountBadgeToken {
        DiscountBadgeToken(
            availableTextColor: Colors.tokenRed100,
            unavailableTextColor: Colors.tokenDark60,
            availableBackgroundColor: Colors.tokenRed100.withAlphaComponent(0.08),
            unavailableBackgroundColor: Colors.tokenDark05,
            actionTextColor: Colors.tokenRed100
        )
    }

    /// Theme mặc định đầy đủ. Đối ứng `PromotionThemeDefaults.defaultConfig(context)` bên Android.
    public static var theme: PromotionSDKTheme {
        PromotionSDKTheme(
            buttonToken: button,
            searchBarToken: searchBar,
            listItemToken: listItem,
            tabChipToken: tabChip,
            tabUnderlineToken: tabUnderline,
            discountBadgeToken: discountBadge
        )
    }

    /// `#AARRGGBB`, rút gọn `#RRGGBB` khi màu đục. Đối ứng `PromotionThemeDefaults.colorToHex`.
    public static func colorToHex(_ color: UIColor) -> String { color.promotionHexString }
}
